### ISSUE TITLE
Why not use existing rust crates for underlying C libs?

### ISSUE THREAD

**Issue by PolyMeilex** [mahkoh/jay#37]
Hi! Sorry that this is an issue, I would rather send a message through a more appropriate channel of communication, but I did not find any.

First of all, this is a really cool project, I'm surprised that I did not notice it earlier. 

So my question is, is there any particular reason why you are not using any of the crates in rust wayland ecosystem (for example libseat-rs, drm-rs, gbm-rs, libinput-rs, wayland-rs, or event Smithay itself). 

Is this just a way to understand the whole stack better?
Or perhaps to be self-contained and minimize a possibility of supply chain attack to basically 0?
Or the one that interest me the most, maybe you just don't like those for x reason, if so I would love to hear about that, and if something could be done to make them usefully for you.
Or perhaps it's something different entirely?

Basically, I'm investigating if maybe work done by Smithay can help you in any way, and reduce effort duplication. If there is something that could be done in this regard, I would gladly work on that.

And once again, sorry for posting totally unrelated stuff to your issue tracker, feel free to delete this.

**Comment by mahkoh** [comment]
> Hi! Sorry that this is an issue, I would rather send a message through a more appropriate channel of communication, but I did not find any.

This is the correct place to ask questions.

> First of all, this is a really cool project, I'm surprised that I did not notice it earlier.

I have not yet officially announced it anywhere and I'm not a big marketer so it is not a surprise that only very few people are aware of it. You are welcome to spread word of it to your friends and colleagues if this project might interest them.

> So my question is, is there any particular reason why you are not using any of the crates in rust wayland ecosystem (for example libseat-rs

I was not aware if this particular crate. It appears to be a rust wrapper for [libseat](https://sr.ht/~kennylevinsen/seatd). I have no interest in using libseat for the following reasons:

- It does not provide an asynchronous API. Currently my code is completely non-blocking.
- It does not provide events for when a device gets removed.
- It does not provide events for when a device gets paused. Instead it only provides very high-level `Enable` and `Disable` events for the seat as a whole.

Therefore I prefer to talk to logind directly.

>, drm-rs, gbm-rs, libinput-rs,

- Creating bindings for drm, libgbm, and libinput is a negligible amount of work in the context of this project. Currently these bindings amount to around 1200 lines while the entire project consists of around 60,000 lines (excluding generated code).
- The rust libraries add a layer of abstraction on top of drm, libgbm, and libinput that are not necessarily the best abstractions for this project. 
- If the rust libraries are missing functionality, I would have to go through upstream to get the necessary features added or I would have to vendor the libraries. This goes against the primary goal of this project to produce a fully functional wayland compositor as fast as possible.

> wayland-rs, or event Smithay itself).

- These libraries have not yet proven themselves. They have been around for a long time but as far as I know they have not yet been used to create a wayland compositor that is used by anyone as their primary desktop environment. Correct me if I'm wrong.
- Smithay is aiming to be a general purpose library for writing wayland compositors. As such it is moving slowly and has to take many competing concerns into account. On the other hand, my compositor is moving extremely quickly and all of its wayland-related code is purpose-built to create a single-threaded tiling wayland compositor designed to use reference counting and interior mutability.
- As with drm-rs etc., Smithay and wayland-rs might be missing support for important features and adding these features would require a lot of work not directly related to my compositor.

> Is this just a way to understand the whole stack better? Or perhaps to be self-contained and minimize a possibility of supply chain attack to basically 0? Or the one that interest me the most, maybe you just don't like those for x reason, if so I would love to hear about that, and if something could be done to make them usefully for you. Or perhaps it's something different entirely?

> Basically, I'm investigating if maybe work done by Smithay can help you in any way, and reduce effort duplication. If there is something that could be done in this regard, I would gladly work on that.

The purpose of the Jay compositor is

- to bring joy to its users (to be honest: user (singular)),
- to provide a tiling wayland compositor,
- to never crash or hang,
- to have all of the nice user-friendly features provided by sway,
- to do all of this as soon as possible.

This project is laser-focused on these goals. I believe that these goals would not be advanced by adding mid-layers between my code and the APIs that I want to actually use. E.g. `jay -> libgbm-rs -> libgbm`, `jay -> wayland-rs -> the wayland wire protocol`.

**Comment by mahkoh** [comment]
> It exposes an `fd` that you can pull, in Smithay's libseat backend we just plug it into our event loop, so it's fully async as well (we don't use async/await syntax, so it may be not obvious). But yeah, this does not matter if you are not interested in libseat in the first place.

I'm talking about opening a device. This is a blocking operation in libseat.

**Comment by PolyMeilex** [comment]
Oh, that's what you mean, if so then yes you are completely right. I did not consider that, as logind backend of Smithay and Weston also blocks on device open, so I assumed the event processing part.

**Comment by mahkoh** [comment]
A few days late but here is a good example: https://github.com/mahkoh/jay/pull/39

This PR implements a new protocol version that is just about getting merged. I saw that is was blocked only on a new implementation and decided to go ahead and implement it. From start to finish, this took me about 10 minutes. 

Getting the example client to compile was in fact significantly more bothersome since I had to copy the new version of the xml file and change the build system to use that xml file.

If Jay did not have its currently level of autarky, I would never have bothered with this since the overhead would have dwarfed the effort to implement the actual business logic.

**Comment by PolyMeilex** [comment]
Indeed, this change is impossible to do without going thought upstreaming process to Smithay or wlroots at some point. Unless you do your own xdg shell implementation.

For new protocols it's easier, one example that comes to my mind is the ext_workspace implementation done by PopOS Cosmic compositor, merging unfinished protocols in Smithay does not make sense, so ext_workspace protocol just lives in tree of cosmic comp, until it is ready to land in wayaland-protocols, and also can be counted as POC server side implementation of the protocol in the MR process of wayland-protocols.


Totally unrelated question, just by reading the code I did not find any mention of libwayland, so the question that instantly comes up is how do you talk to egl/mesa? Do you expose a C api for that? Or there is no EGL buffer support? We have rust implementation of the protocols, but we never figured out a nice and stable way to use it in tandem with egl, so this looks promising.
